src/App.spec.ts

```diff
@@ -94,198 +163,329 @@ describe('App', () => {
-      assert.instanceOf(spyErrorHandler.firstCall.args[0], Error);
-      assert.propertyVal(spyErrorHandler.firstCall.args[0], 'code', ErrorCode.AuthorizationError);
-      assert.propertyVal(spyErrorHandler.firstCall.args[0], 'original', rejection);
+
+      // Assert
+      assert(fakeMiddleware.notCalled);
+      assert(fakeLogger.warn.called);
+      assert.instanceOf(fakeErrorHandler.firstCall.args[0], Error);
+      assert.propertyVal(fakeErrorHandler.firstCall.args[0], 'code', ErrorCode.AuthorizationError);
+      assert.propertyVal(fakeErrorHandler.firstCall.args[0], 'original', dummyAuthorizationError);
     });
     describe('global middleware', () => {
-      it('should process receiver events in order or #use', async () => {
-        const { App } = await importApp();
-        const mockReceiver = createMockReceiver();
-        const mockReceiverEvent = createMockReceiverEvent();
-        const spyFirstMiddleware = createSpyMiddleware();
-        const spySecondMiddleware = createSpyMiddleware();
-        const app = new App({ receiver: mockReceiver, authorize: sinon.fake.resolves({ botToken: '', botId: '' }) });
-        app.use(spyFirstMiddleware);
-        app.use(spySecondMiddleware);
-        (mockReceiver as unknown as EventEmitter).emit('message', mockReceiverEvent);
-        await delay();
-        assert((spyFirstMiddleware as SinonSpy).calledOnce);
-        assert((spyFirstMiddleware as SinonSpy).calledBefore(spySecondMiddleware as SinonSpy));
-        assert((spySecondMiddleware as SinonSpy).calledOnce);
+      it('should process receiver events in order of #use', async () => {
+        // Arrange
+        const fakeReceiver = createFakeReceiver();
+        const fakeFirstMiddleware = sinon.fake(noopMiddleware);
+        const fakeSecondMiddleware = sinon.fake(noopMiddleware);
+        const dummyReceiverEvent = createDummyReceiverEvent();
+        const dummyAuthorizationResult = { botToken: '', botId: '' };
+        const overrides = mergeOverrides(
+          withNoopAppMetadata(),
+          withNoopWebClient(),
+          withMemoryStore(sinon.fake()),
+          withConversationContext(sinon.fake.returns(noopMiddleware)),
+        );
+        const App = await importApp(overrides); // tslint:disable-line:variable-name
+
+        // Act
+        const app = new App({ receiver: fakeReceiver, authorize: sinon.fake.resolves(dummyAuthorizationResult) });
+        app.use(fakeFirstMiddleware);
+        app.use(fakeSecondMiddleware);
+        fakeReceiver.emit('message', dummyReceiverEvent);
+        await delay(10);
+
+        // Assert
+        assert(fakeFirstMiddleware.calledOnce);
+        assert(fakeFirstMiddleware.calledBefore(fakeSecondMiddleware));
+        assert(fakeSecondMiddleware.calledOnce);
+      });
+    });
+    describe('middleware and listener arguments', () => {
+      describe('say()', () => {
+        it.skip('should send a message to a channel where the incoming event originates', async () => {
```

Comment:
Did you mean to leave this unfinished?

Reply:
Yup! I’d like to land this PR and then open a new one that can be merged with #175. That next PR will complete this test and add at least one more that reproduces the bug.